File: 2023/01/ts/src/part2.ts
const STR_TO_NUM: Record<string, number> = {
  one: 1,
  two: 2,
  three: 3,
  four: 4,
  five: 5,
  six: 6,
  seven: 7,
  eight: 8,
  nine: 9,
  zero: 0,
  "1": 1,
  "2": 2,
  "3": 3,
  "4": 4,
  "5": 5,
  "6": 6,
  "7": 7,
  "8": 8,
  "9": 9,
  "0": 0,
}

function getSum(input: string): number {
  let sum = 0
  for (let line of input.split("\n")) {
    let firstDigit: number | undefined
    let lastDigit: number | undefined
    while (line.length > 0) {
      for (const key in STR_TO_NUM) {
        if (line.startsWith(key)) {
          const n = STR_TO_NUM[key]
          if (!firstDigit) {
            firstDigit = n
          } else {
            lastDigit = n
          }
          break
        }
      }
      line = line.slice(1)
    }
    if (firstDigit === undefined) {
      continue
    }
    lastDigit ??= firstDigit
    sum += firstDigit * 10 + lastDigit
  }
  return sum
}

const input = await Bun.file(Bun.argv[2]).text()
const sum = getSum(input)
console.log(sum)

File: 2023/01/rust/src/part2.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

const NUMBERS: [&str; 10] = [
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

pub fn get_result(file: &File) -> u32 {
    let reader = BufReader::new(file);
    let mut sum = 0;

    for line in reader.lines() {
        let line = line.expect("failed to read line");
        let mut first_digit: Option<u32> = None;
        let mut last_digit: Option<u32> = None;

        for i in 0..line.len() {
            let rest = &line[i..];
            let mut found = false;
            for j in 0..NUMBERS.len() {
                if rest.starts_with(NUMBERS[j]) {
                    if first_digit.is_none() {
                        first_digit = Some(j as u32);
                    } else {
                        last_digit = Some(j as u32);
                    }
                    found = true;
                    break;
                }
            }
            if found {
                continue;
            }
            let first_char = rest.chars().next();
            if let Some(c) = first_char {
                if c.is_digit(10) {
                    let digit = c.to_digit(10).unwrap();
                    if first_digit.is_none() {
                        first_digit = Some(digit);
                    } else {
                        last_digit = Some(digit);
                    }
                }
            }
        }

        if first_digit.is_none() {
            continue;
        }
        let first_digit = first_digit.unwrap();
        let last_digit = last_digit.unwrap_or(first_digit);
        sum += first_digit * 10 + last_digit;
    }

    sum
}
